File: sdk/typescript/examples/basic.ts
/**
 * Usage test for the TypeScript SDK.
 *
 * Start the server first:
 *   node dist/src/interfaces/a2a.js
 *
 * Then run:
 *   npm --prefix sdk/typescript run build && node sdk/typescript/dist/examples/basic.js
 */

import { T2CClient } from '../src/index.js';

const TODO_PATH = '.intent-sdk/typescript/TODO.patch';
const AUDIT_PATH = '.intent-sdk/typescript/TODO.patch.json';
const RECEIPT_PATH = '.intent-sdk/typescript/TODO.patch.receipt.json';

interface ExampleContext {
  baseUrl: string;
  token: string | null;
  root: string;
}

interface ExtractionArtifacts {
  graph: Awaited<ReturnType<T2CClient['link']>>;
  diagnostics: Awaited<ReturnType<T2CClient['diagnose']>>;
  extractedRecords: number;
}

interface ProposalArtifacts {
  newProposalIds: string[];
  duplicateProposalIds: string[];
  patchHash: string;
}

async function main(): Promise<void> {
  const context = readExampleContext();
  const client = new T2CClient({ baseUrl: context.baseUrl, token: context.token });

  const extraction = await runExtractionFlow(client, context.root);
  console.log(`extracted ${extraction.extractedRecords} records from ${context.root}`);

  const patch = await runProposalFlow(client, context.root, extraction.graph, extraction.diagnostics);
  console.log(`proposal ids: ${patch.newProposalIds.join(',') || '-'}`);
  console.log(`duplicate ids: ${patch.duplicateProposalIds.join(',') || '-'}`);
  console.log(`patch fingerprint: ${patch.patchHash.slice(0, 16)}`);

  await runRealityAndDiff(client, extraction.graph, extraction.diagnostics, context.root);
  await runWorkspaceComparison(client, context.root);

  console.log('OK');
}

main().catch((error: unknown) => {
  console.error('example failed:', error instanceof Error ? error.message : error);
  process.exitCode = 1;
});

function readExampleContext(): ExampleContext {
  return {
    baseUrl: process.env.T2C_A2A_URL ?? 'http://localhost:8787',
    token: process.env.T2C_A2A_TOKEN ?? null,
    root: process.env.T2C_EXAMPLE_ROOT ?? 'examples/backend',
  };
}

async function runExtractionFlow(client: T2CClient, root: string): Promise<ExtractionArtifacts> {
  const agentCard = await client.agentCard();
  console.log('agent skills:', (agentCard.skills as Array<{ id: string }> | undefined)?.map((skill) => skill.id).join(', '));

  const [nl, ast, markdown] = await Promise.all([
    client.extractNl('task.md', root, 'deterministic'),
    client.extractAst(root),
    client.extractMarkdown(root, { markdownMode: 'deterministic' }),
  ]);

  assertExtractionState(nl.audit, 'NL');
  if (markdown.audit?.status !== 'succeeded') {
    throw new Error(`unexpected Markdown audit: ${JSON.stringify(markdown.audit)}`);
  }
  console.log('markdown audit:', markdown.audit.status, markdown.audit.effectiveMode);

  const records = [...nl.records, ...ast.records, ...markdown.records];
  const graph = await client.link(records);
  console.log('graph fingerprint:', graph.fingerprint.slice(0, 16));

  const diagnostics = await client.diagnose(graph);
  console.log('diagnostics:', diagnostics.counts);
  for (const diagnostic of diagnostics.diagnostics.slice(0, 3)) {
    console.log(`  - [${diagnostic.severity}] ${diagnostic.code}: ${diagnostic.title}`);
  }

  return {
    graph,
    diagnostics,
    extractedRecords: records.length,
  };
}

function assertExtractionState(audit: unknown, label: string): void {
  if (
    typeof audit !== 'object'
    || audit === null
    || !('status' in audit)
    || (audit as { status?: string }).status !== 'succeeded'
  ) {
    throw new Error(`unexpected ${label} audit: ${JSON.stringify(audit)}`);
  }
  console.log(`${label} audit: ${(audit as { status: string; effectiveMode?: string }).status} ${(audit as { effectiveMode?: string }).effectiveMode}`);
}

async function runProposalFlow(
  client: T2CClient,
  root: string,
  graph: Awaited<ReturnType<T2CClient['link']>>,
  diagnostics: Awaited<ReturnType<T2CClient['diagnose']>>,
): Promise<ProposalArtifacts> {
  const synthesis = await client.proposeTodo({
    root,
    graph,
    diagnostics,
    mode: 'prefer-llm',
  });

  const rendered = await client.renderTodo({
    root,
    graph,
    diagnostics,
    synthesis,
    todo: 'TODO.md',
    patch: TODO_PATH,
    audit: AUDIT_PATH,
  });

  const artifact = rendered.artifact as { renderedPatchHash: string };
  const patchHash = artifact.renderedPatchHash;
  await client.applyTodo({
    root,
    todo: 'TODO.md',
    patch: TODO_PATH,
    audit: AUDIT_PATH,
    receipt: RECEIPT_PATH,
    actor: 'sdk-typescript',
    approvalHash: patchHash,
  });

  const validation = synthesis.validation as {
    newProposalIds: string[];
    duplicateProposalIds: string[];
  };

  return {
    newProposalIds: validation.newProposalIds,
    duplicateProposalIds: validation.duplicateProposalIds,
    patchHash,
  };
}

async function runRealityAndDiff(
  client: T2CClient,
  graph: Awaited<ReturnType<T2CClient['link']>>,
  diagnostics: Awaited<ReturnType<T2CClient['diagnose']>>,
  root: string,
): Promise<void> {
  const reality = await client.reality(graph, diagnostics, { includeSvg: true, gapsOnly: true });
  console.log('reality markdown lines:', reality.markdown.split('\n').length);
  console.log('reality svg bytes:', reality.svg?.length ?? 0);

  const gitDiff = await client.diffGit({ root, revision: 'HEAD', includeSvg: true });
  console.log(`git diff files: ${gitDiff.diffs.length}, svg bytes: ${gitDiff.svg?.length ?? 0}`);
}

async function runWorkspaceComparison(client: T2CClient, root: string): Promise<void> {
  if (process.env.T2C_COMPARE_WORKSPACE !== '1') {
    return;
  }

  const comparison = await client.compareWorkspace({ root, base: process.env.T2C_COMPARE_BASE ?? 'origin/main' });
  const direction = (comparison.trend as { direction?: string } | undefined)?.direction;
  console.log('workspace trend:', direction);
}

File: sdk/python/examples/basic.py
#!/usr/bin/env python3
"""Usage test for the todo2code Python SDK.

Start the server first:
    node dist/src/interfaces/a2a.js

Then run:
    python3 sdk/python/examples/basic.py
"""

from __future__ import annotations

import os
import sys
from dataclasses import dataclass
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

from todo2code import T2CClient, T2CError  # noqa: E402

DEFAULT_A2A_URL = "http://localhost:8787"
DEFAULT_EXAMPLE_ROOT = "examples/backend"
DEFAULT_COMPARE_BASE = "origin/main"
TODO_PATCH_PATH = ".intent-sdk/python/TODO.patch"
TODO_AUDIT_PATH = ".intent-sdk/python/TODO.patch.json"
TODO_RECEIPT_PATH = ".intent-sdk/python/TODO.patch.receipt.json"


@dataclass(frozen=True)
class ExampleContext:
    base_url: str
    token: str | None
    root: str
    compare_base: str
    compare_workspace: bool


@dataclass(frozen=True)
class ExtractionArtifacts:
    graph: object
    report: object
    record_count: int


@dataclass(frozen=True)
class ProposalArtifacts:
    new_proposal_ids: tuple[str, ...]
    duplicate_proposal_ids: tuple[str, ...]
    patch_hash: str


def main() -> int:
    context = read_example_context()
    client = T2CClient(context.base_url, token=context.token)

    print("health:", client.health())
    run_flow(client, context)
    print("OK")
    return 0


def read_example_context() -> ExampleContext:
    return ExampleContext(
        base_url=os.environ.get("T2C_A2A_URL", DEFAULT_A2A_URL),
        token=os.environ.get("T2C_A2A_TOKEN"),
        root=os.environ.get("T2C_EXAMPLE_ROOT", DEFAULT_EXAMPLE_ROOT),
        compare_base=os.environ.get("T2C_COMPARE_BASE", DEFAULT_COMPARE_BASE),
        compare_workspace=os.environ.get("T2C_COMPARE_WORKSPACE", "0") == "1",
    )


def run_flow(client: T2CClient, context: ExampleContext) -> None:
    card = client.agent_card()
    print("agent skills:", ", ".join(skill["id"] for skill in card.get("skills", ())))

    extraction = run_extraction_flow(client, context.root)
    print(f"extracted {extraction.record_count} records from {context.root}")

    proposal = run_proposal_flow(
        client,
        context.root,
        extraction.graph,
        extraction.report,
    )
    print("proposal ids:", ",".join(proposal.new_proposal_ids) or "-")
    print("duplicate ids:", ",".join(proposal.duplicate_proposal_ids) or "-")
    print("patch fingerprint:", proposal.patch_hash[:16])

    run_reality_and_diff(client, extraction.graph, extraction.report, context.root)
    run_optional_workspace_comparison(
        client,
        context.root,
        context.compare_base,
        context.compare_workspace,
    )


def run_extraction_flow(client: T2CClient, root: str) -> ExtractionArtifacts:
    nl = client.extract_nl_result("task.md", root, nl_mode="deterministic")
    assert_audit_success(nl.audit, "NL", check_mode=True)
    print("NL audit:", nl.audit.get("status"), nl.audit.get("effectiveMode"))

    ast_records = client.extract_ast(root)
    markdown = client.extract_markdown_result(root, markdown_mode="deterministic")
    assert_audit_success(markdown.audit, "Markdown")
    print("markdown audit:", markdown.audit.get("status"), markdown.audit.get("effectiveMode"))

    records = [*nl.records, *ast_records, *markdown.records]
    graph = client.link(records)
    print("graph fingerprint:", graph.fingerprint[:16])
    print("records by source:", graph.stats.get("bySource"))

    report = client.diagnose(graph)
    print("diagnostics:", dict(report.counts))
    for diagnostic in report.diagnostics[:3]:
        print(f"  - [{diagnostic.severity}] {diagnostic.code}: {diagnostic.title}")

    return ExtractionArtifacts(
        graph=graph,
        report=report,
        record_count=len(records),
    )


def assert_audit_success(
    audit: dict[str, object] | None,
    label: str,
    check_mode: bool = False,
) -> None:
    if audit is None or audit.get("status") != "succeeded":
        raise RuntimeError(f"unexpected {label} audit: {audit}")
    if check_mode and audit.get("effectiveMode") != "deterministic":
        raise RuntimeError(f"unexpected {label} mode: {audit.get('effectiveMode')} for {label}")


def run_proposal_flow(
    client: T2CClient,
    root: str,
    graph: object,
    report: object,
) -> ProposalArtifacts:
    synthesis = client.propose_todo({
        "root": root,
        "graph": graph.raw,
        "diagnostics": report.raw,
        "mode": "prefer-llm",
    })
    validation = synthesis.get("validation", {})

    rendered = client.render_todo({
        "root": root,
        "graph": graph.raw,
        "diagnostics": report.raw,
        "synthesis": synthesis,
        "todo": "TODO.md",
        "patch": TODO_PATCH_PATH,
        "audit": TODO_AUDIT_PATH,
    })

    patch_hash = rendered["artifact"]["renderedPatchHash"]
    client.apply_todo({
        "root": root,
        "todo": "TODO.md",
        "patch": TODO_PATCH_PATH,
        "audit": TODO_AUDIT_PATH,
        "receipt": TODO_RECEIPT_PATH,
        "actor": "sdk-python",
        "approvalHash": patch_hash,
    })

    return ProposalArtifacts(
        new_proposal_ids=tuple(validation.get("newProposalIds", ())),
        duplicate_proposal_ids=tuple(validation.get("duplicateProposalIds", ())),
        patch_hash=patch_hash,
    )


def run_reality_and_diff(
    client: T2CClient,
    graph: object,
    report: object,
    root: str,
) -> None:
    reality = client.reality(graph, report, gapsOnly=True, includeSvg=True)
    print("reality svg bytes:", len(reality.get("svg", "")))
    print(reality["markdown"].split("\n")[4])

    git_diff = client.diff_git(root=root, revision="HEAD", includeSvg=True)
    print("git diff files:", len(git_diff.get("diffs", ())))


def run_optional_workspace_comparison(
    client: T2CClient,
    root: str,
    compare_base: str,
    compare_workspace: bool,
) -> None:
    if not compare_workspace:
        return

    comparison = client.compare_workspace(root=root, base=compare_base)
    print("workspace trend:", comparison.get("trend", {}).get("direction"))


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except T2CError as error:
        print(f"example failed: {error} (code {error.code})", file=sys.stderr)
        raise SystemExit(1) from error
